Form Input: Don't use `flex-direction: row-reverse` for checkbox field

t-hamano opened the pull request:
Follow up #63081

## What?

This PR dynamically changes the markup to remove `flex-direction: row-reverse;` from the Form Input block.

## Why?

In #63081, a stylelint rule was added to disallow flex-direction reverse values.

In the Form Input block, `flex-direction: row-reverse;` is applied to place text after the checkbox element. Currently, it is allowed with a disable comment, but ideally, this rule should not be used.

## How?

If the `type` attribute is `checkbox` or `radio`, place the label after the input element. This means that the HTML will be changed to the following:

```html
<!-- From -->
<div class="wp-block-form-input">
	<label class="wp-block-form-input__label">
		<span class="wp-block-form-input__label-content">Label</span>
		<input class="wp-block-form-input__input" type="checkbox" name="label" aria-required="false"/>
	</label>
</div>

<!-- To -->
<div class="wp-block-form-input">
	<label class="wp-block-form-input__label">
		<input class="wp-block-form-input__input" type="checkbox" name="label" aria-required="false"/>
		<span class="wp-block-form-input__label-content">Label</span>
	</label>
</div>
```

Also added block deprecation since we are updating the `save` function.

Note: Updating the HTML for the Form Input block was once attempted in #63081, but was ultimately removed due to the need for block deprecation. The code in this PR is based on [this commit](https://github.com/WordPress/gutenberg/pull/63081/commits/db35911bb0445c5de8291304032f34609d72014e).

## Testing Instructions

- Switch to the trunk branch.
- Through the code editor, insert the following HTML and save the post. This is a single Form Input block, and the input element is a checkbox:
  ```html
  <!-- wp:form-input {"type":"checkbox"} -->
  <div class="wp-block-form-input">
  	<label class="wp-block-form-input__label">
  		<span class="wp-block-form-input__label-content">Label</span>
  		<input class="wp-block-form-input__input" type="checkbox" name="label" aria-required="false"/>
  	</label>
  </div>
  <!-- /wp:form-input -->
  ```
- On the front end, make sure the checkbox is ordered first, then the label.
- Switch to the `form-input/remove-row-reverse` branch.
- Reload the browser on the front end.
- The order of the checkbox and the label should remain unchanged.
- Open the post editor, update the checkbox label, and save the post. The markup will be updated as follows:
  ```html
  <!-- wp:form-input {"type":"checkbox"} -->
  <div class="wp-block-form-input">
  	<label class="wp-block-form-input__label">
  		<input class="wp-block-form-input__input" type="checkbox" name="label" aria-required="false"/>
  		<span class="wp-block-form-input__label-content">Label</span>
  	</label>
  </div>
  <!-- /wp:form-input -->
  ```
- Reload the browser on the front end.
- The order of the checkbox and the label should remain unchanged.

t-hamano committed:
Form Input: Don't use `flex-direction: row-reverse` for checkbox field

t-hamano commented on test/integration/fixtures/blocks/core__form-input__deprecated-v2.serialized.html:
```diff
@@ -0,0 +1,3 @@
+<!-- wp:form-input {"type":"checkbox","className":"wp-block-form-input"} -->
```
Block deprecation works fine, but one thing that bothers me is that for some reason the block base class name is applied as the custom class name 🤔

![image](https://github.com/user-attachments/assets/6d7f78a3-3223-4162-a6fe-2d9f396b8b85)

aaronrobertshaw commented on packages/block-library/src/form-input/deprecated.js:
```diff
@@ -26,6 +27,121 @@ const getNameFromLabelV1 = ( content ) => {
+		},
+	},
+	supports: {
+		className: false,
```
```suggestion
```

Trunk doesn't disable className support. This might be the cause of the bug noted in https://github.com/WordPress/gutenberg/pull/64232#discussion_r1703116726

aaronrobertshaw commented on packages/block-library/src/form-input/deprecated.js:
```diff
@@ -26,6 +27,121 @@ const getNameFromLabelV1 = ( content ) => {
+					<TagName
+						className={ inputClasses }
+						type={ 'textarea' === type ? undefined : type }
+						name={ name || getNameFromLabelV2( label ) }
```
```suggestion
						name={ name || getNameFromLabelV1( label ) }
```

I don't think we need to create a new frozen util definition when it hasn't changed at all.

t-hamano replied:
Fixed by 946fdc1

aaronrobertshaw commented on packages/block-library/src/form-input/deprecated.js:
```diff
@@ -26,6 +27,121 @@ const getNameFromLabelV1 = ( content ) => {
 	);
 };
 
+const getNameFromLabelV2 = getNameFromLabelV1;
```
```suggestion
```

I don't think this is necessary.

The util hasn't changed since it was frozen for the first deprecation. You can read the util's name here as the `V1` for `getNameFromLabel`.

t-hamano replied:
Fixed by 946fdc1

aaronrobertshaw reviewed (commented):
Thanks for the PR @t-hamano 👍 

After manually testing this one, it is mostly working for me with the exception of the bug you noted with the default block class name being saved as a custom class name. I think that might be related to a misconfiguration of the deprecation `supports`.

I've left some minor inline comments relating to the new deprecation. 

The last thing I noticed is that the original deprecation's fixtures weren't updated. All deprecations have to bring that version of the block up to the latest markup. 

This should mean the original `v1` deprecation's serialized HTML fixture needed an update. The fact that it didn't and passes might indicate there's another issue here.

aristath commented:
I think this might be related to https://github.com/WordPress/gutenberg/pull/55755 as well...

t-hamano committed:
Don't create v2 variable

t-hamano committed:
Fix fixtures

t-hamano commented:
Thanks for the review!

> Trunk doesn't disable className support. This might be the cause of the bug noted in https://github.com/WordPress/gutenberg/pull/64232#discussion_r1703116726

This was probably the cause 😅 I removed this and the fixture was generated as expected.

> The last thing I noticed is that the original deprecation's fixtures weren't updated. All deprecations have to bring that version of the block up to the latest markup.

This PR only updates markup if the type attribute is `radio` or `checkbox`. But the `type` attribute of the `v1` fixture is `text`. I think this is the reason why the serialized v1 fixture is not updated.

aaronrobertshaw commented:
> This PR only updates markup if the type attribute is radio or checkbox. But the type attribute of the v1 fixture is text. I think this is the reason why the serialized v1 fixture is not updated.

Thanks for the explanation, it probably means we're missing some fixtures to cover all bases for the block.

t-hamano committed:
Update fixtures

aaronrobertshaw reviewed (commented):
Thanks for the quick work adding the extra fixtures @t-hamano 🚀 

I have a couple of suggestions for the new fixtures though. 

Like other unit tests, limiting the scope of each test to a single use case is generally beneficial. Other existing blocks limit their fixtures to a single block configuration per fixture e.g. image custom link, media link, captions etc. all getting their own fixture. The different types of form input likely warrant their own independent fixture.

Another benefit in structuring fixtures this way is that it's easier to find relevant fixtures to update when the next deprecation needs to be implemented.

While breaking down the updated v1 fixtures, it might be better to add base fixtures for the block rather than include them under the v1 fixtures. For example: `core__form-input__radio`, `core__form-input__checkbox` etc..

This gives a little more flexibility for deprecation fixtures to only target block configurations that are impacted by the relevant deprecation.

What do you think?

t-hamano committed:
Split fixtures

t-hamano commented:
Sounds like a good idea!

I've split it into [all available types as variations](https://github.com/WordPress/gutenberg/blob/form-input/remove-row-reverse/packages/block-library/src/form-input/variations.js), but should we remove fixtures that aren't affected by the deprecation?

aaronrobertshaw commented:
> I've split it into [all available types as variations](https://github.com/WordPress/gutenberg/blob/form-input/remove-row-reverse/packages/block-library/src/form-input/variations.js), but should we remove fixtures that aren't affected by the deprecation?

Sorry I wasn't more clear with my suggestion @t-hamano.

I was only suggesting that we had a default or base fixture for each of the different form input types. The deprecation fixtures can be specific to the use case or type they are relevant for e.g. the v2 deprecation fixture could only be a checkbox and/or radio as that's the main focus of the change.

Some of the input types are just a change of attribute on the `input` element so maybe those those don't need their own fixture e.g. `url`, `tel` etc.

I'm probably missing some but maybe we just need base fixtures for:
- `core__form-input.html` (text type, which also pretty much covers `url`, `tel` etc)
- `core__form-input__checkbox.html`
- `core__form-input__radio.html`
- `core__form-input__textarea.html`

Does that help clarify things? If not I'm happy to try again 😅

t-hamano committed:
Integrate fixtures

t-hamano commented:
I reorganized it into the following four fixtures.

Are there any fixtures that we don't need? Sorry if I didn't understand 😅

- `core__form-input__deprecated-v1.html`
- `core__form-input__deprecated-v2.html`
- `core__form-input__radio__deprecated-v1.html`
- `core__form-input__radio__deprecated-v2.html`
- `core__form-input__checkbox__deprecated-v1.html`
- `core__form-input__checkbox__deprecated-v2.html`
- `core__form-input__textarea__deprecated-v1.html`
- `core__form-input__textarea__deprecated-v2.html`

aaronrobertshaw commented:
> Sorry if I didn't understand 😅

It's probably more that I'm not explaining the fixtures I'm proposing very well, without listing every file out individually.

> Are there any fixtures that we don't need? 

After the latest change, we're still missing the base fixtures for the different input types as outlined in https://github.com/WordPress/gutenberg/pull/64232#issuecomment-2292850432.

> core__form-input.html (text type, which also pretty much covers url, tel etc)
core__form-input__checkbox.html
core__form-input__radio.html
core__form-input__textarea.html

These cover the latest (current) version of the block. 

The v2 deprecation was only for checkbox and radio inputs, correct? So the deprecation fixtures for v2 probably don't need ones for textarea or other input types that aren't effected. This what I meant by this comment:

> The deprecation fixtures can be specific to the use case or type they are relevant for e.g. the v2 deprecation fixture could only be a checkbox and/or radio as that's the main focus of the change.

We don't have to follow my recommendation if you don't agree with it. If you do, what I think is left is to:

1. Create base fixtures for a generic text input, checkbox, radio, and textarea. These are _not_ fixtures for deprecations, just the current block version.
2. Remove the deprecation fixtures for input types unrelated to the deprecation. If a block's markup isn't changed from the base fixture, it's already covered for the deprecation.

If this isn't making things clearer, I can carve out some time early next week to create a diff or push a commit here if it helps.

t-hamano committed:
Remove unneccesary fixtures

t-hamano commented:
I removed the fixtures that I deemed unnecessary. The current fixtures consist of:

- **v1**: This deprecation is required for all form element types because it adds the useBlocksProps to the save function ([#56507](https://github.com/WordPress/gutenberg/pull/56507))
  - `core__form-input__deprecated-v1.html`: Includes default, `email`, `number`, `tel`, `url` types
  - `core__form-input__radio__deprecated-v1.html`
  - `core__form-input__checkbox__deprecated-v1.html`
  - `core__form-input__textarea__deprecated-v1.html`
- **v2**: This is the target of this PR, swapping the input and label elements. This deprecation only affects the `radio` and `checkbox` types.
  - `core__form-input__radio__deprecated-v2.html`
  - `core__form-input__checkbox__deprecated-v2.html`

If I'm still misunderstanding, feel free to push 🙏

aaronrobertshaw commented:
Thanks @t-hamano

> 1. Create base fixtures for a generic text input, checkbox, radio, and textarea. These are not fixtures for deprecations, just the current block version.

⬆️ This is the bit still missing  

> If I'm still misunderstanding, feel free to push 🙏

Unfortunately, I might not be able to get to this before I'm AFK until the 10th September.

t-hamano commented:
@aaronrobertshaw Thanks for the reply.

This PR is not something that needs to be rushed, so I would be happy if you could take a look at it again when you have time.

aaronrobertshaw committed:
Add base fixtures for text-like inputs

All variations of the simple texts input field are added here to match the approach taken in the deprecation fixtures.

aaronrobertshaw committed:
Add base fixtures for checkbox, radio, and textarea inputs

aaronrobertshaw commented:
@t-hamano I've rebased this and pushed some base fixtures for the various form input types. I've also aligned the default fixture with how you combined all the text, email, number, tel, url inputs in the deprecation fixtures.

Given I've contributed the last couple of commits, I'd like to get a fresh set of eyes on this PR for final approval.

@andrewserong if you can spare a moment to review this, that would be greatly appreciated 🙏 

From my end, I think this should be good to go!

andrewserong reviewed (commented):
Thanks for the ping!

In testing this locally I noticed that the removal of the CSS rule means that any unmigrated content that is already on a site will continue to display in the other order until going in to a post and updating / saving so that the migration is run.

The deprecation runs correctly in the editor:

![image](https://github.com/user-attachments/assets/80ccfab1-f92c-4471-a80e-913e8492bd7c)

However prior to running that deprecation in the editor, this is how requests to the site frontend looked with this PR applied:

<img src="https://github.com/user-attachments/assets/db3efd53-d3e9-40d8-893e-9c6e81e87117" width="250" />

After adding a paragraph to the post and saving, the update takes effect as expected:

<img src="https://github.com/user-attachments/assets/54b4b5ec-2641-4fdd-99e1-96dc0818fdc4" width="250" />

I see this block is still gated behind an experimental feature flag. Does that mean it's safe to make this change because no-one's using it in production sites? Or do we still need to handle markup that's already out in the wild and hasn't been migrated via the deprecation?

Aside from that, code-wise this change LGTM.

t-hamano commented:
@aaronrobertshaw Thanks for the update! By looking at your commits, I was able to understand your intentions.

> In testing this locally I noticed that the removal of the CSS rule means that any unmigrated content that is already on a site will continue to display in the other order until going in to a post and updating / saving so that the migration is run.

This is likely to occur for all blocks whose save function has been updated.

I know some approaches that leave CSS in place for backward compatibility, but I don't know of any clear standards for how much old markup should be supported with CSS.

Personally, I feel that for form blocks, it's enough to just ensure that the block doesn't break.

If we want to add CSS that takes old markup into account, you can use the following approach:

```css
// The old markup, i.e. if there is an input element after the label content, reverse the order as before.
.wp-block-form-input__label:has(.wp-block-form-input__label-content + input[type="checkbox"]) {
	flex-direction: row-reverse;
}
```

andrewserong commented:
> This is likely to occur for all blocks whose save function has been updated.
> I know some approaches that leave CSS in place for backward compatibility, but I don't know of any clear standards for how much old markup should be supported with CSS.

From my perspective, if it's a block that's used in production, it probably either needs CSS that we keep around pretty much forever (as we can't guarantee that old post content on big sites will ever be migrated), _or_ use a server-side approach to dynamically update old markup when the block is rendered.

For the Form blocks, since it's behind an experiment, I imagine it's not so important because the blocks aren't being used in production anywhere? I'm not certain of this assumption, though, so I'll just ping @WordPress/gutenberg-core for visibility on that question.

ntsekouras commented:
>From my perspective, if it's a block that's used in production, it probably either needs CSS that we keep around pretty much forever (as we can't guarantee that old post content on big sites will ever be migrated), or use a server-side approach to dynamically update old markup when the block is rendered.

Yes, we should keep even the CSS for changes in blocks for back compat.

>Personally, I feel that for form blocks, it's enough to just ensure that the block doesn't break.

I'd echo this, since it's an experiment.. Also if the css rule is something really simple and would keep the old version as is, I don't see why not add it.

t-hamano committed:
Merge branch 'trunk' into form-input/remove-row-reverse

t-hamano committed:
Add fallback style for old markup

t-hamano commented:
Thank you for the feedback, everyone! I've added some CSS for the old markup.

I've also updated the Testing Instructions section so you can test it.

ciampo commented:
@t-hamano , I was reading the PR description and got confused by the "From" -> "To" code snippet in the "How" section, since the two snippets are identical. Is that a typo? We should update it so that folks visiting this PR can have a better comprehension of its changes :) Thank you!!

t-hamano commented:
@ciampo Thanks for letting me know! That was definitely a typo. I updated it correctly.

aaronrobertshaw reviewed (approved):
Thanks for the patience and iteration here @t-hamano. I definitely appreciate this became more involved than expected.

✅ Deprecated version continues to render correctly on the frontend prior to migration
✅ Migration of deprecated version in the editor works as advertised
✅ Migrated block appears correctly on frontend once saved
✅ Deprecation and block fixtures look good and all pass

LGTM 🚢

andrewserong commented:
> Thanks for the patience and iteration here

+1 — my apologies, I meant to come back to this PR and it totally slipped my mind. Thanks for all the follow-up!

t-hamano commented:
Thank you everyone for the reviews 🙇‍♂️
